File: frontend/src/lib/api/tickets.ts
import { z } from 'zod';

// Mirrors TicketResult from POST /api/tickets — the filed Linear issue's human
// identifier (e.g. "MY-42") and its url.
export const ticketResultSchema = z.object({
	identifier: z.string(),
	url: z.string()
});

export type TicketResult = z.infer<typeof ticketResultSchema>;

// Mirrors TicketCreate (snake_case on the wire). `page_url` records where the user
// was when they filed it.
export type TicketInput = {
	title: string;
	description: string;
	page_url: string | null;
};

/** Whether the in-app ticket form is offered — true only when the backend's Linear
 *  integration is configured. `fetchFn` is injectable for SSR/tests. */
export async function fetchTicketsEnabled(fetchFn: typeof fetch = fetch): Promise<boolean> {
	const res = await fetchFn('/api/tickets/enabled');
	if (!res.ok) throw new Error(`GET /api/tickets/enabled → ${res.status}`);
	const body = (await res.json()) as { enabled?: unknown };
	return body.enabled === true;
}

/** File a ticket, opening an issue in the Cookmarks Linear project. Rejects if the
 *  backend couldn't file it (e.g. Linear unreachable). `fetchFn` is injectable. */
export async function submitTicket(
	input: TicketInput,
	fetchFn: typeof fetch = fetch
): Promise<TicketResult> {
	const res = await fetchFn('/api/tickets', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(input)
	});
	if (!res.ok) throw new Error(`POST /api/tickets → ${res.status}`);
	return ticketResultSchema.parse(await res.json());
}

File: backend/app/api/tickets.py
import logging

from fastapi import APIRouter, HTTPException

from app.schemas.ticket import TicketCreate, TicketResult
from app.services.linear import LinearError, create_issue, linear_configured

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/tickets", tags=["tickets"])


def _build_description(ticket: TicketCreate) -> str:
    parts = [ticket.description.strip()] if ticket.description.strip() else []
    if ticket.page_url:
        parts.append(f"\n\n---\n**Reported from:** {ticket.page_url}")
    return "\n".join(parts) or "_No description provided._"


@router.get("/enabled")
def tickets_enabled() -> dict[str, bool]:
    """Whether the in-app ticket form should be offered — true only when the Linear
    integration is configured, so the footer link stays hidden otherwise."""
    return {"enabled": linear_configured()}


@router.post("", response_model=TicketResult, status_code=201)
def submit_ticket(ticket: TicketCreate) -> TicketResult:
    """File a ticket as an issue in the Cookmarks Linear project. Returns 503 if Linear
    is unreachable or unconfigured, so the UI can tell the user it didn't go through."""
    title = ticket.title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="A ticket title is required")
    try:
        issue = create_issue(title, _build_description(ticket))
    except LinearError as exc:
        logger.warning("Ticket not filed: %s", exc)
        raise HTTPException(status_code=503, detail=str(exc)) from exc

    return TicketResult(identifier=issue["identifier"], url=issue["url"])
